Improve performance of semi-automatic modeling assessments

SiggZ opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] ~I added integration test cases for the client (Jest) related to the features~
- [x] ~I added screenshots/screencast of my UI changes~
- [x] ~I translated all the newly inserted strings~

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
The new automatic assessment mechanism implemented in https://github.com/ls1intum/Artemis/pull/548 has some major performance issues when there are a lot of modeling submissions without assessment. Especially, when a manual assessment gets submitted by a tutor, it has a high impact on users. The automatic assessment of all submissions is triggered after a manual assessment is submitted which results in a response time of ~25-30 seconds after a submit (for 1000 existing submissions in the database).

### Description
<!-- Describe your changes in detail -->
To cope with this problem, we change the storage of the automatic results. After an automatic result is created, it will not get stored to the database anymore. Instead, a ConcurrentHashMap will be used in Compass to store the automatic results. As soon as a submission is locked for assessment, the corresponding automatic result will be retrieved from the hash map and stored to the database. Additionally, we adjust the process of calculating the next optimal submissions that should be assessed next. 

The changes in detail:
- Store automatic results in hash map instead of database to reduce the number of database calls.
- Retrieve automatic result from hash map and store it together with submission to database when a submission is locked for assessment.
- Remove automatic result from hash map after it is stored to the database (and after an assessment is canceled).
- Load submission together with result, feedback, assessor, participation and participation results in one database access to reduce the number of database calls.
- Load all submissions without assessment from the database at once when the automatic assessment of all submissions is triggered (e.g. after a manual assessment submit) instead of loading each submission in an a separate iteration.
- Optimal submissions: When the number of cached optimal submissions is smaller than a configured threshold of 10 submissions, 10 new optimal submissions are calculated by the `ModelSelector` and loaded into the cache, instead of just loading one new optimal submission every time. By that, the user has to wait for the calculation only every 10th assessment on average.
- Optimal submissions: Every time 10 new optimal submissions are calculated by the `ModelSelector`, they are calculated at once, instead of doing the calculation 10 times in a row.
- Optimal submissions: If Compass could assign automatic feedback to all the elements of each submission, 10 random submissions are now chosen, as there are no optimal submissions anymore if everything is already assessed automatically. Also, there was a small "bug" in this case that lead to the situation that every submission without an assessment was compared to every other submission without assessment resulting in tremendous calculation effort. Selecting a new optimal submission took about 12 seconds for ~1000 existing submissions. By fixing this we reduced the calculation time to ~1 second.

Through these changes the following performance improvements could be achieved:
- The response time after submitting an assessment decreased from 25-30 seconds to ~1.5 seconds.
- The time of initializing Compass (after a server restart) decreased from over 5 minutes to ~4.5 second.
- The time of loading new submissions into the next optimal submission cache takes ~1 second, now 10 submissions are loaded at once instead of only one.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Precondition: a high number of submissions (~1000) that are not assessed exist in the database for one exercise.

1. Log in to ArTEMiS as a tutor
2. Navigate to the tutor exercise dashboard of the corresponding exercise
3. Submit an assessment for a submission
4. Check response time

krusche commented:
Are the changes already tested on the test server? Can we merge this?

krusche commented on src/main/java/de/tum/in/www1/artemis/service/compass/CompassService.java:
```diff
@@ -53,6 +54,13 @@
+     * Map submissionIds to automatic results. Automatic results generated by Compass are not stored in the database, instead they are stored in this map. As soon as a submission
+     * is locked for assessment, the corresponding automatic result will be retrieved from this map and stored to the database.
+     */
+    // TODO CZ: this map contains the automatic results for the submissions of all exercises. Should we create a separate automatic result map for the different exercises?
```
Please have a look how we handle this in quizzes. I probably makes sense to split this into multiple ConcurrentHashMaps per exercise, because this could decrease the changes of problems due to concurrent access

SiggZ replied:
I changed the `Map<Long, Result>` to a `Map<Long, Map<Long, Result>>` that separates the automatic results by exercuse,

krusche commented on src/main/java/de/tum/in/www1/artemis/service/compass/CompassService.java:
```diff
@@ -237,72 +257,159 @@ public void addAssessment(long exerciseId, long submissionId, List<Feedback> mod
+     * @param submissionId the id of the modeling submission for which an automatic result should be generated/updated
+     * @param exerciseId   the id of the corresponding exercise
+     */
+    private void assessAutomatically(long submissionId, long exerciseId) {
```
Why is this method even invoked for unsupported modeling exercises? Can we prevent this earlier?

SiggZ replied:
I added the `isSupported` check to all relevant public methods of the `CompassService` and removed it from the private methods.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/compass/CompassService.java:
```diff
@@ -237,72 +257,159 @@ public void addAssessment(long exerciseId, long submissionId, List<Feedback> mod
-    private void assessAutomatically(long modelId, long exerciseId) {
+    private void assessAllAutomatically(Collection<Long> submissionIds, long exerciseId) {
+        // Workaround for ignoring automatic assessments of unsupported modeling exercise types TODO remove this after adapting compass
+        if (!isSupported(exerciseId)) {
```
see above

SiggZ replied:
I added the `isSupported` check to all relevant public methods of the `CompassService` and removed it from the private methods.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/compass/controller/ModelSelector.java:
```diff
@@ -2,13 +2,16 @@
 
     private static final int MAX_CANDIDATE_LIST_SIZE = 50;
 
+    private static final double EPSILON = 0.0000001;
```
What is this used for? Could you shortly document both constants?

SiggZ replied:
done.

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/repository/ModelingSubmissionRepository.java:
```diff
@@ -23,8 +25,26 @@
-    @Query("select distinct submission from ModelingSubmission submission left join fetch submission.result left join fetch submission.participation p left join fetch p.results where submission.id = :#{#submissionId}")
-    Optional<ModelingSubmission> findByIdWithEagerResultAndParticipationResults(@Param("submissionId") Long submissionId);
+    /**
+     * Load the modeling submission with the given id together with its result, the feedback list of the result, the assessor of the result, its participation and all results of
```
Very good documentation!

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingAssessmentResource.java:
```diff
@@ -286,4 +297,19 @@ public ResponseEntity cancelAssessment(@PathVariable Long submissionId) {
+     * @param result        the result that should be sent to the client
+     * @return true if the user is allowed to retrieve the given result, false otherwise
+     */
+    private boolean isUserAllowedToGetAssessment(Exercise exercise, Participation participation, Result result) {
```
Would it be possible to move this to the AuthorizationCheckService?
Also there is a similiar method to `courseService.userHasAtLeastStudentPermissions(exercise.getCourse())` in AuthorizationCheckService.isStudentInCourse.

I sometimes wonder if we should try to centralize the permissions more into one place to enforce reuse  :thinking:.

SiggZ replied:
You are right, I moved the method to the `AuthorizationCheckService`.

SiggZ commented:
Thank you for reviewing. I adjusted the PR and addressed your comments. I also tested the functionality on the test server now and it looks good to me. However, the Travis builds fail now due to the memory issue: `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`.
